File: core/model/src/main/kotlin/dev/fluentmai/android/core/model/PlayerRecords.kt
package dev.fluentmai.android.core.model

import kotlin.math.absoluteValue
import kotlin.math.roundToLong

/** Stable public chart identity. Room row ids are deliberately not part of it. */
data class ChartIdentity(
    val songId: Int,
    val songType: SongType,
    val difficulty: Difficulty,
) {
    fun stableKey(): String = "$songId:${songType.name}:${difficulty.name}"

    companion object {
        fun from(chart: ChartRecord): ChartIdentity =
            ChartIdentity(chart.songId, chart.songType, chart.difficulty)

        fun from(score: ScoreRecord): ChartIdentity? =
            score.songId?.takeIf { it > 0 }?.let {
                ChartIdentity(it, score.songType, score.difficulty)
            }

        fun parseStableKey(value: String?): ChartIdentity? {
            val parts = value?.split(':') ?: return null
            if (parts.size != 3) return null
            val songId = parts[0].toIntOrNull()?.takeIf { it > 0 } ?: return null
            val songType = SongType.entries.firstOrNull { it.name == parts[1] } ?: return null
            val difficulty = Difficulty.entries.firstOrNull { it.name == parts[2] } ?: return null
            return ChartIdentity(songId, songType, difficulty)
        }
    }
}

enum class AchievementRank(val displayName: String) {
    SSS_PLUS("SSS+"),
    SSS("SSS"),
    SS_PLUS("SS+"),
    SS("SS"),
    S_PLUS("S+"),
    S("S"),
    AAA("AAA"),
    AA("AA"),
    A("A"),
    BBB("BBB"),
    BB("BB"),
    B("B"),
    C("C"),
    D("D");

    companion object {
        fun fromAchievement(value: Double): AchievementRank =
            when {
                value >= 100.5 -> SSS_PLUS
                value >= 100.0 -> SSS
                value >= 99.5 -> SS_PLUS
                value >= 99.0 -> SS
                value >= 98.0 -> S_PLUS
                value >= 97.0 -> S
                value >= 94.0 -> AAA
                value >= 90.0 -> AA
                value >= 80.0 -> A
                value >= 75.0 -> BBB
                value >= 70.0 -> BB
                value >= 60.0 -> B
                value >= 50.0 -> C
                else -> D
            }
    }
}

enum class FullComboStatus(val displayName: String) {
    FC("FC"),
    FC_PLUS("FC+"),
    AP("AP"),
    AP_PLUS("AP+"),
    UNKNOWN("未知");

    val satisfiesFullCombo: Boolean
        get() = this != UNKNOWN
    val satisfiesAllPerfect: Boolean
        get() = this == AP || this == AP_PLUS

    companion object {
        fun fromWireValue(value: String?): FullComboStatus? =
            when (normalizeFlag(value)) {
                "fc" -> FC
                "fcp", "fcplus" -> FC_PLUS
                "ap" -> AP
                "app", "applus" -> AP_PLUS
                "" -> null
                else -> UNKNOWN
            }
    }
}

enum class FullSyncStatus(val displayName: String) {
    SYNC("SYNC"),
    FS("FS"),
    FS_PLUS("FS+"),
    FSD("FSD"),
    FSD_PLUS("FSD+"),
    UNKNOWN("未知");

    val satisfiesFullSyncDx: Boolean
        get() = this == FSD || this == FSD_PLUS

    companion object {
        fun fromWireValue(value: String?): FullSyncStatus? =
            when (normalizeFlag(value)) {
                "sync" -> SYNC
                "fs" -> FS
                "fsp", "fsplus" -> FS_PLUS
                "fsd", "fdx" -> FSD
                "fsdp", "fdxp", "fsdplus", "fdxplus" -> FSD_PLUS
                "" -> null
                else -> UNKNOWN
            }
    }
}

data class PlayerChartRecord(
    val identity: ChartIdentity,
    val chart: ChartRecord,
    val score: ScoreRecord?,
    val rating: Int?,
) {
    val rank: AchievementRank?
        get() = score?.let { AchievementRank.fromAchievement(it.achievement) }
    val fullComboStatus: FullComboStatus?
        get() = FullComboStatus.fromWireValue(score?.fc)
    val fullSyncStatus: FullSyncStatus?
        get() = FullSyncStatus.fromWireValue(score?.fs)
}

data class PlayerRecordCatalog(
    val records: List<PlayerChartRecord>,
    val unmatchedScoreCount: Int,
)

data class PlayerRecordStats(
    val totalCharts: Int,
    val playedCharts: Int,
    val unplayedCharts: Int,
    val unmatchedScores: Int,
    val rankCounts: Map<AchievementRank, Int>,
    val fullComboCounts: Map<FullComboStatus, Int>,
    val fullSyncCounts: Map<FullSyncStatus, Int>,
)

/**
 * Matches persisted scores to public chart identities without ever using a Room row id.
 * Title fallback is deliberately limited to a single public identity so same-title songs
 * cannot silently borrow each other's chart metadata.
 */
fun matchChartsForScores(
    charts: List<ChartRecord>,
    scores: List<ScoreRecord>,
): Map<String, ChartRecord> {
    val chartByIdentity = charts.distinctBy(ChartIdentity::from).associateBy(ChartIdentity::from)
    val uniqueChartByFallback = charts
        .groupBy { chart ->
            ScoreFallbackKey(normalizePlayerRecordText(chart.title), chart.songType, chart.levelIndex)
        }
        .mapValues { (_, matches) -> matches.distinctBy(ChartIdentity::from) }
        .filterValues { it.size == 1 }
        .mapValues { (_, matches) -> matches.single() }
    return scores.mapNotNull { score ->
        val direct = ChartIdentity.from(score)?.let(chartByIdentity::get)
        val fallback = uniqueChartByFallback[
            ScoreFallbackKey(normalizePlayerRecordText(score.title), score.songType, score.levelIndex)
        ]
        (direct ?: fallback)?.let { score.id to it }
    }.toMap()
}

fun buildPlayerRecordCatalog(
    charts: List<ChartRecord>,
    scores: List<ScoreRecord>,
): PlayerRecordCatalog {
    val bestByIdentity = scores
        .mapNotNull { score -> ChartIdentity.from(score)?.let { it to score } }
        .bestScorePairs()
    val bestByTitle = scores.bestScoresBy { score ->
        ScoreFallbackKey(normalizePlayerRecordText(score.title), score.songType, score.levelIndex)
    }
    val chartIdentities = charts.map(ChartIdentity::from).toSet()
    val chartIdentitiesByFallbackKey = charts.groupBy { chart ->
        ScoreFallbackKey(normalizePlayerRecordText(chart.title), chart.songType, chart.levelIndex)
    }.mapValues { (_, matchingCharts) -> matchingCharts.map(ChartIdentity::from).distinct() }
    val uniqueChartFallbackKeys = chartIdentitiesByFallbackKey
        .filterValues { identities -> identities.size == 1 }
        .keys
    val chartFallbackKeys = chartIdentitiesByFallbackKey.keys
    val records = charts
        .distinctBy(ChartIdentity::from)
        .map { chart ->
            val identity = ChartIdentity.from(chart)
            val fallback = ScoreFallbackKey(
                normalizePlayerRecordText(chart.title),
                chart.songType,
                chart.levelIndex,
            )
            val score = bestByIdentity[identity]
                ?: bestByTitle[fallback]?.takeIf { fallback in uniqueChartFallbackKeys }
            PlayerChartRecord(
                identity = identity,
                chart = chart,
                score = score,
                rating = score?.let { chart.levelValue?.let { constant -> calculateDxRating(constant, score.achievement, score.fc) } },
            )
        }
    val unmatchedScoreCount = scores.count { score ->
        val identityMatched = ChartIdentity.from(score)?.let { it in chartIdentities } == true
        val fallback = ScoreFallbackKey(
            normalizePlayerRecordText(score.title),
            score.songType,
            score.levelIndex,
        )
        val fallbackMatched = fallback in chartFallbackKeys && fallback in uniqueChartFallbackKeys
        !identityMatched && !fallbackMatched
    }
    return PlayerRecordCatalog(records, unmatchedScoreCount)
}

fun PlayerRecordCatalog.stats(records: List<PlayerChartRecord> = this.records): PlayerRecordStats =
    PlayerRecordStats(
        totalCharts = records.size,
        playedCharts = records.count { it.score != null },
        unplayedCharts = records.count { it.score == null },
        unmatchedScores = unmatchedScoreCount,
        rankCounts = records.mapNotNull { it.rank }.groupingBy { it }.eachCount(),
        fullComboCounts = records.mapNotNull { it.fullComboStatus }.groupingBy { it }.eachCount(),
        fullSyncCounts = records.mapNotNull { it.fullSyncStatus }.groupingBy { it }.eachCount(),
    )

enum class VersionAgeFilter { ALL, CURRENT, OLD }

enum class PlateKind(val displayName: String) {
    GENERAL("将"),
    EXTREME("极"),
    GOD("神"),
    MAIMAI("舞舞"),
    CONQUEROR("霸者"),
}

data class PlateBlocker(
    val record: PlayerChartRecord,
    val currentValue: String,
    val requirementGap: String,
)

data class PlateProgress(
    val kind: PlateKind,
    val versionId: Int?,
    val versionName: String,
    val requiredCount: Int,
    val completedCount: Int,
    val blockers: List<PlateBlocker>,
    val eligibleRecords: List<PlayerChartRecord>,
    val dataSufficient: Boolean,
    val dataMessage: String? = null,
    val plateName: String = if (kind == PlateKind.CONQUEROR) "覇者" else "$versionName${kind.displayName}",
) {
    val remainingCount: Int = (requiredCount - completedCount).coerceAtLeast(0)
    val completionFraction: Double = if (requiredCount == 0) 0.0 else completedCount.toDouble() / requiredCount
    val isComplete: Boolean = dataSufficient && requiredCount > 0 && remainingCount == 0
}

/**
 * Rules follow SEGA's official 2020-01-15 plate announcement.
 * Version plates use BASIC through MASTER; Re:MASTER is excluded. 覇者 uses every
 * STANDARD chart from BASIC through Re:MASTER and the official 80% clear threshold.
 */
fun calculatePlateProgress(
    records: List<PlayerChartRecord>,
    kind: PlateKind,
    versionId: Int?,
    versionName: String?,
): PlateProgress {
    if (kind != PlateKind.CONQUEROR && versionId == null) {
        return PlateProgress(kind, null, versionName.orEmpty(), 0, 0, emptyList(), emptyList(), false, "缺少版本信息")
    }
    val plateVersion = versionId?.let(::maimaiPlateVersionFor)
    if (kind != PlateKind.CONQUEROR && plateVersion == null) {
        return PlateProgress(
            kind,
            versionId,
            versionName.orEmpty(),
            0,
            0,
            emptyList(),
            emptyList(),
            false,
            "该曲库版本还没有可核验的版本牌要求",
        )
    }
    if (kind != PlateKind.CONQUEROR && plateVersion?.supports(kind) != true) {
        return PlateProgress(
            kind,
            versionId,
            versionName.orEmpty(),
            0,
            0,
            emptyList(),
            emptyList(),
            false,
            "该版本没有${kind.displayName}牌要求",
        )
    }
    val eligible = records.filter { record ->
        when (kind) {
            PlateKind.CONQUEROR -> record.chart.songType == SongType.STANDARD
            else -> plateVersion?.contains(record.chart) == true && record.chart.difficulty != Difficulty.RE_MASTER
        }
    }
    if (eligible.isEmpty()) {
        return PlateProgress(
            kind,
            versionId,
            versionName.orEmpty(),
            0,
            0,
            emptyList(),
            emptyList(),
            false,
            if (kind == PlateKind.CONQUEROR) "曲库中没有可核验的标准谱面" else "该版本没有可核验的 BASIC～MASTER 谱面",
        )
    }
    val blockers = eligible.mapNotNull { record -> record.plateBlocker(kind) }
    return PlateProgress(
        kind = kind,
        versionId = if (kind == PlateKind.CONQUEROR) null else versionId,
        versionName = if (kind == PlateKind.CONQUEROR) "全标准谱面" else versionName.orEmpty(),
        requiredCount = eligible.size,
        completedCount = eligible.size - blockers.size,
        blockers = blockers,
        eligibleRecords = eligible,
        dataSufficient = true,
        plateName = if (kind == PlateKind.CONQUEROR) {
            "覇者"
        } else {
            checkNotNull(plateVersion).displayTitle(kind)
        },
    )
}

private fun PlayerChartRecord.plateBlocker(kind: PlateKind): PlateBlocker? {
    val score = score
    return when (kind) {
        PlateKind.GENERAL -> {
            val achievement = score?.achievement ?: 0.0
            if (achievement >= 100.0) null else PlateBlocker(
                this,
                score?.let { "${it.achievement.fixedFourDecimals()}%" } ?: "未游玩",
                "达成率还差 ${(100.0 - achievement).coerceAtLeast(0.0).fixedFourDecimals()}%",
            )
        }
        PlateKind.EXTREME -> if (fullComboStatus?.satisfiesFullCombo == true) null else PlateBlocker(
            this,
            fullComboStatus?.displayName ?: "未达成",
            "需要 FC 或更高状态",
        )
        PlateKind.GOD -> if (fullComboStatus?.satisfiesAllPerfect == true) null else PlateBlocker(
            this,
            fullComboStatus?.displayName ?: "未达成",
            "需要 AP 或 AP+",
        )
        PlateKind.MAIMAI -> if (fullSyncStatus?.satisfiesFullSyncDx == true) null else PlateBlocker(
            this,
            fullSyncStatus?.displayName ?: "未达成",
            "需要 FSD 或 FSD+",
        )
        PlateKind.CONQUEROR -> {
            val achievement = score?.achievement ?: 0.0
            if (achievement >= 80.0) null else PlateBlocker(
                this,
                score?.let { "${it.achievement.fixedFourDecimals()}%" } ?: "未游玩",
                "距 CLEAR 还差 ${(80.0 - achievement).coerceAtLeast(0.0).fixedFourDecimals()}%",
            )
        }
    }
}

private fun Double.fixedFourDecimals(): String {
    val scaled = (this * 10_000.0).roundToLong()
    val sign = if (scaled < 0L) "-" else ""
    val magnitude = scaled.absoluteValue
    return "$sign${magnitude / 10_000}.${(magnitude % 10_000).toString().padStart(4, '0')}"
}

fun PlayerChartRecord.isEligibleForPlate(kind: PlateKind): Boolean =
    when (kind) {
        PlateKind.CONQUEROR -> chart.songType == SongType.STANDARD
        else -> chart.difficulty != Difficulty.RE_MASTER
    }

fun PlayerChartRecord.meetsPlateRequirement(kind: PlateKind): Boolean = plateBlocker(kind) == null

private data class ScoreFallbackKey(
    val title: String,
    val songType: SongType,
    val levelIndex: Int,
)

private fun normalizePlayerRecordText(value: String): String =
    normalizeUnicodeCompatibility(value.trim())
        .lowercase()
        .replace(Regex("[\\s._·・:：!！?？'\"“”‘’()（）\\[\\]【】/\\\\-]+"), "")

private fun normalizeFlag(value: String?): String =
    value.orEmpty().trim().lowercase()
        .replace("+", "plus")
        .replace("_", "")
        .replace("-", "")
        .replace(" ", "")

private fun <K> List<ScoreRecord>.bestScoresBy(keySelector: (ScoreRecord) -> K): Map<K, ScoreRecord> =
    groupBy(keySelector).mapValues { (_, values) -> values.bestScore() }

private fun <K> List<Pair<K, ScoreRecord>>.bestScorePairs(): Map<K, ScoreRecord> =
    groupBy({ it.first }, { it.second }).mapValues { (_, values) -> values.bestScore() }

private fun List<ScoreRecord>.bestScore(): ScoreRecord =
    maxWithOrNull(compareBy<ScoreRecord> { it.achievement }.thenBy { it.dxScore ?: -1 }) ?: first()

File: windows/fluentmai_core/player_records.py
from __future__ import annotations

from collections import Counter
from dataclasses import dataclass
from enum import Enum

from .chart_browser import ChartRecord
from .version_catalog import plate_version_for, version_name_for


class AchievementRank(str, Enum):
    SSS_PLUS = "SSS+"
    SSS = "SSS"
    SS_PLUS = "SS+"
    SS = "SS"
    S_PLUS = "S+"
    S = "S"
    AAA = "AAA"
    AA = "AA"
    A = "A"
    BBB = "BBB"
    BB = "BB"
    B = "B"
    C = "C"
    D = "D"


class FullComboStatus(str, Enum):
    FC = "FC"
    FC_PLUS = "FC+"
    AP = "AP"
    AP_PLUS = "AP+"
    UNKNOWN = "未知"


class FullSyncStatus(str, Enum):
    SYNC = "SYNC"
    FS = "FS"
    FS_PLUS = "FS+"
    FSD = "FSD"
    FSD_PLUS = "FSD+"
    UNKNOWN = "未知"


class PlateKind(str, Enum):
    GENERAL = "general"
    EXTREME = "extreme"
    GOD = "god"
    MAIMAI = "maimai"
    CONQUEROR = "conqueror"

    @property
    def display_name(self) -> str:
        return {
            PlateKind.GENERAL: "将",
            PlateKind.EXTREME: "极",
            PlateKind.GOD: "神",
            PlateKind.MAIMAI: "舞舞",
            PlateKind.CONQUEROR: "霸者",
        }[self]


@dataclass(frozen=True)
class PlayerStats:
    total_charts: int
    played_charts: int
    unplayed_charts: int
    rank_counts: dict[AchievementRank, int]
    full_combo_counts: dict[FullComboStatus, int]
    full_sync_counts: dict[FullSyncStatus, int]


@dataclass(frozen=True)
class PlateBlocker:
    record: ChartRecord
    current_value: str
    requirement_gap: str


@dataclass(frozen=True)
class PlateProgress:
    kind: PlateKind
    version_id: int | None
    version_name: str
    plate_name: str
    required_count: int
    completed_count: int
    blockers: tuple[PlateBlocker, ...]
    eligible_records: tuple[ChartRecord, ...]
    data_sufficient: bool
    data_message: str | None = None

    @property
    def remaining_count(self) -> int:
        return max(0, self.required_count - self.completed_count)

    @property
    def is_complete(self) -> bool:
        return self.data_sufficient and self.required_count > 0 and self.remaining_count == 0


def achievement_rank(value: float) -> AchievementRank:
    for threshold, rank in (
        (100.5, AchievementRank.SSS_PLUS), (100.0, AchievementRank.SSS),
        (99.5, AchievementRank.SS_PLUS), (99.0, AchievementRank.SS),
        (98.0, AchievementRank.S_PLUS), (97.0, AchievementRank.S),
        (94.0, AchievementRank.AAA), (90.0, AchievementRank.AA),
        (80.0, AchievementRank.A), (75.0, AchievementRank.BBB),
        (70.0, AchievementRank.BB), (60.0, AchievementRank.B),
        (50.0, AchievementRank.C),
    ):
        if value >= threshold:
            return rank
    return AchievementRank.D


def full_combo_status(value: str | None) -> FullComboStatus | None:
    normalized = _normalize_flag(value)
    if not normalized:
        return None
    return {
        "fc": FullComboStatus.FC,
        "fcp": FullComboStatus.FC_PLUS,
        "fcplus": FullComboStatus.FC_PLUS,
        "ap": FullComboStatus.AP,
        "app": FullComboStatus.AP_PLUS,
        "applus": FullComboStatus.AP_PLUS,
    }.get(normalized, FullComboStatus.UNKNOWN)


def full_sync_status(value: str | None) -> FullSyncStatus | None:
    normalized = _normalize_flag(value)
    if not normalized:
        return None
    return {
        "sync": FullSyncStatus.SYNC,
        "fs": FullSyncStatus.FS,
        "fsp": FullSyncStatus.FS_PLUS,
        "fsplus": FullSyncStatus.FS_PLUS,
        "fsd": FullSyncStatus.FSD,
        "fdx": FullSyncStatus.FSD,
        "fsdp": FullSyncStatus.FSD_PLUS,
        "fdxp": FullSyncStatus.FSD_PLUS,
        "fsdplus": FullSyncStatus.FSD_PLUS,
        "fdxplus": FullSyncStatus.FSD_PLUS,
    }.get(normalized, FullSyncStatus.UNKNOWN)


def player_stats(records: list[ChartRecord]) -> PlayerStats:
    played = [record for record in records if record.played]
    return PlayerStats(
        total_charts=len(records),
        played_charts=len(played),
        unplayed_charts=len(records) - len(played),
        rank_counts=dict(Counter(achievement_rank(record.achievements or 0.0) for record in played)),
        full_combo_counts=dict(Counter(filter(None, (full_combo_status(record.full_combo) for record in played)))),
        full_sync_counts=dict(Counter(filter(None, (full_sync_status(record.full_sync) for record in played)))),
    )


def calculate_plate_progress(
    records: list[ChartRecord],
    kind: PlateKind,
    version_id: int | None = None,
) -> PlateProgress:
    version_name = version_name_for(version_id)
    plate = plate_version_for(version_id)
    if kind != PlateKind.CONQUEROR and version_id is None:
        return _unavailable(kind, version_id, version_name, "缺少版本信息")
    if kind != PlateKind.CONQUEROR and plate is None:
        return _unavailable(kind, version_id, version_name, "该曲库版本还没有可核验的版本牌要求")
    if kind != PlateKind.CONQUEROR and kind.value not in plate.supported_kinds:
        return _unavailable(kind, version_id, version_name, f"该版本没有{kind.display_name}牌要求")

    if kind == PlateKind.CONQUEROR:
        eligible = [record for record in records if record.chart_type == "SD" and not record.is_utage]
        plate_name = "霸者"
        label = "全标准谱面"
    else:
        assert plate is not None
        eligible = [
            record for record in records
            if record.difficulty_index != 4
            and not record.is_utage
            and record.chart_version is not None
            and plate.contains(record.song_id, record.chart_version)
        ]
        plate_name = " / ".join(f"{prefix}{kind.display_name}" for prefix in plate.prefixes)
        label = version_name
    if not eligible:
        message = "曲库中没有可核验的标准谱面" if kind == PlateKind.CONQUEROR else "该版本没有可核验的 BASIC～MASTER 谱面"
        return _unavailable(kind, version_id, label, message, plate_name=plate_name)

    blockers = tuple(
        blocker for record in eligible if (blocker := _plate_blocker(record, kind)) is not None
    )
    return PlateProgress(
        kind=kind,
        version_id=None if kind == PlateKind.CONQUEROR else version_id,
        version_name=label,
        plate_name=plate_name,
        required_count=len(eligible),
        completed_count=len(eligible) - len(blockers),
        blockers=blockers,
        eligible_records=tuple(eligible),
        data_sufficient=True,
    )


def _plate_blocker(record: ChartRecord, kind: PlateKind) -> PlateBlocker | None:
    achievement = record.achievements or 0.0
    combo = full_combo_status(record.full_combo)
    sync = full_sync_status(record.full_sync)
    if kind == PlateKind.GENERAL:
        return None if achievement >= 100.0 else PlateBlocker(record, _achievement_text(record), f"达成率还差 {max(0.0, 100.0 - achievement):.4f}%")
    if kind == PlateKind.EXTREME:
        return None if combo not in {None, FullComboStatus.UNKNOWN} else PlateBlocker(record, combo.value if combo else "未达成", "需要 FC 或更高状态")
    if kind == PlateKind.GOD:
        return None if combo in {FullComboStatus.AP, FullComboStatus.AP_PLUS} else PlateBlocker(record, combo.value if combo else "未达成", "需要 AP 或 AP+")
    if kind == PlateKind.MAIMAI:
        return None if sync in {FullSyncStatus.FSD, FullSyncStatus.FSD_PLUS} else PlateBlocker(record, sync.value if sync else "未达成", "需要 FSD 或 FSD+")
    return None if achievement >= 80.0 else PlateBlocker(record, _achievement_text(record), f"距 CLEAR 还差 {max(0.0, 80.0 - achievement):.4f}%")


def _achievement_text(record: ChartRecord) -> str:
    return f"{record.achievements:.4f}%" if record.played else "未游玩"


def _unavailable(kind, version_id, version_name, message, *, plate_name=None):
    return PlateProgress(kind, version_id, version_name, plate_name or ("霸者" if kind == PlateKind.CONQUEROR else f"{version_name}{kind.display_name}"), 0, 0, (), (), False, message)


def _normalize_flag(value: str | None) -> str:
    return (value or "").strip().lower().replace("+", "plus").replace("_", "").replace("-", "").replace(" ", "")
